// @tarojs/mini-runner/dist/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GLOBAL_PROPS = exports.excludeReplaceTaroFrameworkPkgs = exports.PARSE_AST_TYPE = exports.QUICKAPP_SPECIAL_COMPONENTS = void 0;
const helper_1 = require("@tarojs/helper");
exports.QUICKAPP_SPECIAL_COMPONENTS = new Set([
    'View',
    'Text',
    'Block'
]);
var PARSE_AST_TYPE;
(function (PARSE_AST_TYPE) {
    PARSE_AST_TYPE["ENTRY"] = "ENTRY";
    PARSE_AST_TYPE["PAGE"] = "PAGE";
    PARSE_AST_TYPE["COMPONENT"] = "COMPONENT";
    PARSE_AST_TYPE["NORMAL"] = "NORMAL";
    PARSE_AST_TYPE["STATIC"] = "STATIC";
    PARSE_AST_TYPE["EXPORTS"] = "EXPORTS";
})(PARSE_AST_TYPE = exports.PARSE_AST_TYPE || (exports.PARSE_AST_TYPE = {}));
exports.excludeReplaceTaroFrameworkPkgs = new Set([helper_1.taroJsRedux, helper_1.taroJsMobx, helper_1.taroJsMobxCommon]);
exports.GLOBAL_PROPS = '{Function: Function,Boolean: Boolean,Object: Object,Number: Number,Array: Array,Date: Date,String: String,Symbol: Symbol,Error: Error,TypeError: TypeError,Map: Map,Set: Set,WeakMap: WeakMap,WeakSet: WeakSet,ArrayBuffer: ArrayBuffer,Math: Math,Promise: Promise,RegExp: RegExp,DataView: DataView,isFinite: isFinite,parseInt: parseInt,parseFloat: parseFloat,Float32Array: Float32Array,Float64Array: Float64Array,Int8Array: Int8Array,Int16Array: Int16Array,Int32Array: Int32Array,Uint8Array: Uint8Array,Uint16Array: Uint16Array,Uint32Array: Uint32Array,Uint8ClampedArray: Uint8ClampedArray,setTimeout: setTimeout,clearTimeout: clearTimeout,setInterval: setInterval,clearInterval: clearInterval}';
